### ISSUE TITLE
ConsumerGroupMetadata is internal and there is no (clear) way for an external library to wrap it

### ISSUE THREAD

**Issue by agayev169** [confluentinc/confluent-kafka-dotnet#1529]
`IConsumerGroupMetadata` is an empty interface, that does not have a public property/method to access the metadata itself, while `ConsumerGroupMetadata` has an internal field containing the serialized version of metadata. 

https://github.com/confluentinc/confluent-kafka-dotnet/blob/2b54dd2ab07f6eaa886d88c57143838627b2512f/src/Confluent.Kafka/ConsumerGroupMetadata.cs#L23

Due to this design, the only way to get metadata of a consumer group is to cast `IConsumerGroupMetadata` to `ConsumerGroupMetadata`.

https://github.com/confluentinc/confluent-kafka-dotnet/blob/2b54dd2ab07f6eaa886d88c57143838627b2512f/src/Confluent.Kafka/Impl/SafeKafkaHandle.cs#L553

A better way to do it would be to have a public property to get metadata as `byte[]` in `IConsumerGroupMetadata` itself.
Please, consider this change.

P.S. I will try to do these changes in my free time, but maybe there is a samaritan that could do this now.

**Comment by mhowlett** [comment]
serializedMetadata is an internal format that isn't intended to be publicly visible.

you are right that it would be good for ConsumerGroupMetadata to expose something useful, but that will be a bigger task. probably a flaw in how we did this tbh (though nothing that can't be worked around).

**Comment by edenhill** [comment]
@agayev169 The ConsumerGroupMetadata's purpose is to pass consumer group state from a consumer to a transactional producer, the state information is of no use to the application and thus there's no reason to expose it.

**Comment by mhowlett** [comment]
the java client actually exposes information on the instance returned from `groupMetadata`, which is what is passed to the transactional producer. whether on not this is really useful (seems like @agayev169 want's it), our api does expose less than the Java one does.

**Comment by agayev169** [comment]
> @agayev169 The ConsumerGroupMetadata's purpose is to pass consumer group state from a consumer to a transactional producer, the state information is of no use to the application and thus there's no reason to expose it.

I get your point, and partially agree with you on this point; however, let's consider my case. I am trying to develop a service using Kafka. I do not want to leave Kafka dependencies in the public API of the service (maybe it will be change in the future, who knows), so I tried to wrap all the components that I need under my own public API (consumer, producer, message, etc.). The only problem that I faced till now is that I cannot wrap the `ICosumerGroupMetadata` since it does not let me get the metadata itself from my service (both `ConsumerGroupMetadata` and variable storing metadata are internal).

So, what I find missing is a public method that would let me get the metadata as `byte[]`.

**Comment by mhowlett** [comment]
i *think* that is a valid use case, though i'd need to think the problem through to be sure. i don't expect it's very common though. i'm also wary of putting a method `Serialize` and constructor on this class, because it may imply it makes sense for it to be persisted.

can i suggest using reflection to get at the internal byte array and to create a new ConsumerGroupMetadata instance? we don't guarantee the internals won't change, but I don't think they will, and if they do, worst case is they will change to something that still allows you to achieve what you want, but you'll need to update your code.

**Comment by agayev169** [comment]
To be honest, I do not like the idea of using reflection in this case - it feels really hacky.
I think having a public API is only an advantage because it will only let you *get* the value from an external library, not *set* it. I will have a look at the code again to satisfy your point about Serialization.

**Comment by mhowlett** [comment]
yes, of course it's a hack. but your users won't know about it and it means we don't need to commit to an addition to the API which I feel has downsides, and we may want to back out if we change what is exposed from this class.

**Comment by agayev169** [comment]
> yes, of course it's a hack. but your users won't know about it and it means we don't need to commit to an addition to the API which I feel has downsides, and we may want to back out if we change what is exposed from this class.

My main concern is the speed of execution. The service I am developing is going to receive many requests, so the request handling should be as fast as possible.

**Comment by edenhill** [comment]
The ConsumerGroupMetadata only needs to be acquired once per transaction (to send offsets), so performance should not be a concern.

**Comment by agayev169** [comment]
> The ConsumerGroupMetadata only needs to be acquired once per transaction (to send offsets), so performance should not be a concern.

Sure, but the point is that my transactions should be as fast as possible, and creating the ConsumerGroupMetadata with reflections may harm the overall performance. However, I will try to do it and what is the situation.
